doc/sphinx/src/tasks.rst

```diff
@@ -120,21 +116,52 @@ is shown below.
 .. figure:: figs/TaskDiagram.png
    :alt: Task Diagram
 
-``TaskCollection`` provides two member functions, ``AddRegion`` and
-``Execute``.
-
-AddRegion
-~~~~~~~~~
-
-``AddRegion`` simply adds a new ``TaskRegion`` to the back of the
-collection and returns it as a reference. The integer argument
-determines how many task lists make up the region.
-
-Execute
-~~~~~~~
-
-Calling the ``Execute`` method on the ``TaskCollection`` executes all
-the tasks that have been added to the collection, processing each
-``TaskRegion`` in the order they were added, and allowing tasks in
-different ``TaskList``\ s but the same ``TaskRegion`` to be executed
-concurrently.
+``TaskCollection`` provides a few 
+public-facing functions:
+- ``TaskRegion& AddRegion(const int num_lists)``: Add and return a reference to
+a new ``TaskRegion`` with the specified number of ``TaskList``s.
+- ``TaskListStatus Execute(ThreadPool &pool)``: Execute all regions in the
+collection.  Regions are executed completely, in the order they were added,
+before moving on to the next region.  Task execution will take advantage of
+the provided ``ThreadPool`` to (possibly) execute tasks across ``TaskList``s
+in each region concurrently.
+- ``TaskListStatus Execute()``: Same as above, but execution will use an
+internally generated ``ThreadPool`` with a single thread.
+
+NOTE: Work remains to make the rest of
+Parthenon thread-safe, so it is currently not recommended to use a ``ThreadPool``
+with more than one thread.
+
+TaskQualifier
+-------------
+
+``TaskQualifier``s provide a mechanism for downstream codes to alter the default
+behavior of specific tasks in certain ways.  The qualifiers are described below:
+- ``TaskQualifier::local_sync``: Tasks marked with ``local_sync`` synchronize across
+lists in a region on a given MPI rank.  Tasks that depend on a ``local_sync``
+marked task gain dependencies from the corresponding task on all lists within
+a region.  A typical use for this qualifier is to do a rank-local reduction, for
+example before initiating a global MPI reduction (which should be done only once
+per rank, not once per ``TaskList``).  Note that Parthenon links tasks across
+lists in the order they are added to each list, i.e. the ``n``th ``local_sync`` task
+in a list is assumed to be associated with the ``n``th ``local_sync`` task in all
+lists in the region.
+- ``TaskQualifier::global_sync``: Tasks marked with ``global_sync`` implicitly have
+the same semantics as ``local_sync``, but additionally do a global reduction on the
```

Comment:
I think so.   The important distinction of `global_sync` is that it does a reduction on the statuses across all MPI ranks to determine if a task is complete or not.  Technically this may not be required if all MPI ranks are guaranteed to return the same status, as would be the case if they all checked some condition on some globally reduced quantity.  Given that, we can probably avoid using `global_sync` in some places where one might naively think that a reduction on statuses is required.  But I can imagine use cases where different ranks evaluate different things and the global reduction of statuses would be required.

Reply:
My thought is that if we want to sync all ranks, there is always going to be some information that gets communicated between them (more than just that they are done). Maybe I am not being imaginative enough though.

Reply:
let me try to make something up quickly.  Say we're pushing particles around that can move across many cells in a given time step.  One algorithm you could code up would have every rank push all the particles that live on it's blocks until the end of the time step or until it leaves the rank's domain (at which point it sends them to the right neighbor).  You could write this as an iterative thing with the completion check being a check of whether all particles on the rank have reached the end of the time step.  Then the completion check is totally local to the rank, but you need to continue iterating globally until everybody agrees things are done.

Reply:
Sure, but that could also just be written as a reduction to get the total number of active particles in the simulation.